plugins/vbe/fu-vbe-simple-device.c

```diff
@@ -0,0 +1,844 @@
+	     prio++, entry = g_list_next(entry)) {
+		const gchar *compat, *p, *pend;
+		const gchar *cmp = entry->data;
+		gint ret, len;
+
+		compat = fdt_getprop(fit->blob, cfg, "compatible", &len);
+		g_debug("compat compare with: %s", cmp);
+		if (!compat) {
+			g_debug("   (none)");
+		} else {
+			for (p = compat, pend = compat + len; p < pend; p += strlen(p) + 1)
+				g_debug("   %s", p);
+		}
+		ret = fdt_node_check_compatible(fit->blob, cfg, cmp);
+		if (!ret || ret == -FDT_ERR_NOTFOUND)
+			return prio;
+	}
+
+	return -1;
+}
+
+/**
+ * process_image() - Write a single image to the device
+ *
+ * Look up an image node, read the data out of it and write the data to the
+ * device at the correct store offset.
+ *
+ * @fit: FIT to write
+ * @img: FIT offset of image to write
+ * @self: Device to use
+ * @error: Returns error information if this function returns FALSE
+ * Returns: TRUE on success, FALSE on failure
+ */
+static gboolean
+process_image(FuVbeSimpleDevice *self, struct fit_info *fit, gint img, GError **error)
+{
+	guint store_offset = 0;
+	const gchar *buf;
+	off_t seek_to;
+	gint size;
+	gint ret;
+
+	ret = fit_img_store_offset(fit, img);
+	if (ret >= 0) {
+		store_offset = ret;
+	} else if (ret != -FITE_NOT_FOUND) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "Image '%s' store offset is invalid (%d)",
+			    fit_img_name(fit, img),
+			    ret);
+		return FALSE;
+	}
+
+	buf = fit_img_data(fit, img, &size);
+	if (buf == NULL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "Image '%s' data could not be read (%d)",
+			    fit_img_name(fit, img),
+			    size);
+		return FALSE;
+	}
+
+	if (store_offset + size > self->area_size) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "Image '%s' store_offset=%#x, size=%#x, area_size=%#jx",
+			    fit_img_name(fit, img),
+			    (guint)store_offset,
+			    (guint)size,
+			    (uintmax_t)self->area_size);
+		return FALSE;
+	}
+
+	if (self->skip_offset >= size) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "Image '%s' skip_offset=%#x, size=%#x, area_size=%#jx",
+			    fit_img_name(fit, img),
+			    (guint)store_offset,
+			    (guint)size,
+			    (uintmax_t)self->area_size);
+		return FALSE;
+	}
+
+	seek_to = self->area_start + store_offset + self->skip_offset;
+	g_debug("Writing image '%s' size 0x%x (skipping 0x%x) to store_offset 0x%x, seek %jx\n",
+		fit_img_name(fit, img),
+		(guint)size,
+		(guint)self->skip_offset,
+		store_offset,
+		(uintmax_t)seek_to);
+
+	ret = lseek(self->fd, seek_to, SEEK_SET);
+	if (ret < 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "Cannot seek file '%s' (%d) to %#jx (%s)",
+			    self->devname,
+			    self->fd,
+			    (uintmax_t)seek_to,
+			    strerror(errno));
+		return FALSE;
+	}
+
+	ret = write(self->fd, buf + self->skip_offset, size - self->skip_offset);
+	if (ret < 0) {
```

Comment:
@sjg20 does this take a significant amount of time? Could we write it in lots of chunks so that we can advance the FuProgress bar in a smooth way or does the write have to be done all at once?